### PR TITLE
New Release 3.1.1

### PR DESCRIPTION
**Pull request by petretiandrea**

### CODE REVIEW

File: custom_components/tapo/light.py [hunk 1]
```diff
@@ -191,15 +204,17 @@ async def _change_brightness(self, new_brightness, apply_to_effect: str = None):
...
-    color_modes = [ColorMode.ONOFF]
+    color_modes = []
     if device.is_color_temperature:
         color_modes.append(ColorMode.COLOR_TEMP)
-    if device.is_brightness:
-        color_modes.append(ColorMode.BRIGHTNESS)
     if device.is_color:
         color_modes.append(ColorMode.HS)
+    if device.is_brightness and not device.is_color_temperature and not device.is_color:
+        color_modes.append(ColorMode.BRIGHTNESS)
```
Comment (bramstroker) [inline custom_components/tapo/light.py:211]: I think you should just return here `{ColorMode.BRIGHTNESS}` and not append to the set, as the documentation states it should be the only possible color mode.
Same applies to `ONOFF`.

Screenshot 2024-04-13 at 09 56 38

So for example supported modes set to `color_temp, brightness` should never be possible.

Comment (petretiandrea) [reply custom_components/tapo/light.py:211]: The code follows the HA algorithm describe here https://developers.home-assistant.io/docs/core/entity/light/#color-modes. Don't you agree?

Comment (bramstroker) [reply custom_components/tapo/light.py:211]: That's also the resource where my screenshot is from.
So this is not correct, as in the current implementation you have now there could potentially be scenarios where `brightness` could be mixed with other color modes in the `supported_color_modes`, but it has to be the only one.

See also the explanation how to correctly determine the color modes:

Screenshot 2024-04-13 at 12 43 09

Comment (petretiandrea) [reply custom_components/tapo/light.py:211]: The `if` has the `and not device.is_color_temperature and not device.is_color` so cannot be mixed with other color modes. No?

Comment (bramstroker) [reply custom_components/tapo/light.py:211]: Yes, I see you are correct about that.
However I see potential issue with the last condition:

```
if not device.is_color_temperature and not device.is_color:
        color_modes.append(ColorMode.ONOFF)
```

This does not check for `brightness`.
So you'll end up with `brightness, onoff` as supported color modes, which is not valid.

Maybe rewrite to:

```python
if not color_modes:
     color_modes.append(ColorMode.ONOFF)
```

Similarly as HA suggests in their approach.

Comment (petretiandrea) [reply custom_components/tapo/light.py:211]: Ok, now It's clear and I agree with you

**Pull request merged**